uniffi_bindgen/src/bindings/backend.rs

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::interface::*;
+use std::fmt;
+
+pub type TypeIdentifier = Type;
+pub type Literal = crate::interface::Literal;
+
+/// An object to look up a foreign language code specific renderer for a given type used.
+/// Every `Type` referred to in the `ComponentInterface` should map to a corresponding
+/// `CodeType`.
+///
+/// The mapping may be opaque, but the oracle always knows the answer.
+///
+/// In adddition, the oracle knows how to render identifiers (function names,
+/// class names, variable names etc).
+pub trait CodeOracle {
+    fn find(&self, type_: &TypeIdentifier) -> Box<dyn CodeType>;
+
+    /// Get the idiomatic rendering of a class name (for enums, records, errors, etc).
+    fn class_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of a function name.
+    fn fn_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of a variable name.
+    fn var_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of an individual enum variant.
+    fn enum_variant_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of an exception name
+    ///
+    /// This replaces "Error" at the end of the name with "Exception".  Rust code typically uses
+    /// "Error" for any type of error but in the Java world, "Error" means a non-recoverable error
+    /// and is distinguished from an "Exception".
+    fn error_name(&self, nm: &dyn fmt::Display) -> String;
+
+    fn ffi_type_label(&self, ffi_type: &FFIType) -> String;
+}
+
+/// A Trait to emit foreign language code to handle referenced types.
+/// A type which is specified in the UDL (i.e. a member of the component interface)
+/// will have a `CodeDeclaration` as well, but for types used e.g. primitive types, Strings, etc
+/// only a `CodeType` is needed.
+pub trait CodeType {
+    /// The language specific label used to reference this type. This will be used in
+    /// method signatures and property declarations.
+    fn type_label(&self, oracle: &dyn CodeOracle) -> String;
+
+    /// A representation of this type label that can be used as part of another
+    /// identifier. e.g. `read_foo()`, or `FooInternals`.
+    ///
+    /// This is especially useful when creating specialized objects or methods to deal
+    /// with this type only.
+    fn canonical_name(&self, oracle: &dyn CodeOracle) -> String {
+        self.type_label(oracle)
+    }
+
+    /// A representation of the given literal for this type.
+    /// N.B. `Literal` is aliased from `interface::Literal`, so may not be whole suited to this task.
+    fn literal(&self, oracle: &dyn CodeOracle, literal: &Literal) -> String;
```

Comment:
Should literal be a `CodeType` method?  This implies that we would handle literals differently depending on which type they're being assigned to.  It seems simpler to just have 1 literal function and have the target language figure out if the assignment is correct or not.

Reply:
IIRC `Literal`s are already typed, either implicitly or as part of the enum.

The agreement between the argument type and the literal is type is checked in `uniffi_bindgen/src/interface/literal.rs`.

In general, I think the backends should need to do as validity checking/error detection on the `ComponentInterface` as possible.

Reply:
But your question: should it be a method in `CodeType`? 

We do get some additional benefits around checking that the type and literals match up, but I don't think it's the point.

One of the motivations for this refactor was to reduce the need for huge exhaustive `match` expressions on `Type`. 

The analogous GOF pattern this would be refactoring to would be the State pattern, which has one `match` expression to decide what state you're in and then the state be able to answer all the specific questions you were matching on before.

The motivation for that is two fold: 

1. make it easier to add and remove states.
2. make it easier to reason about states in isolation.

By leaving the literal function as a mega-`match` makes it slightly harder to do each of these.

Reply:
Gosh darn it. "as _little_ validity checking". Yes.